LedgerSMB/Database/Change.pm

```diff
@@ -0,0 +1,280 @@
+
+
+
+=head2 content_wrapped($before, $after)
+
+Wrap a file with more statements in the same transaction.
+
+So you get something like:
+
+BEGIN;
+$before
+$self->content
+$after
+COMMIT;
+
+Useful for db updates so you can update version numbers or the like.
+
+=cut
+
+sub content_wrapped {
+    my ($self, $before, $after) = @_;
+    $before //= "";
+    $after //= "";
+    return $self->_wrap_transaction(
+        _wrap($self->content(1), $before, $after)
+    );
+}
+
+=head2 is_applied($dbh)
+
+Returns true if the current sha matches one that has been applied.
+
+=cut
+
+sub is_applied {
+    my ($self, $dbh) = @_;
+    my $sha = $self->sha;
+    my $sth = $dbh->prepare(
+        "SELECT * FROM db_patches WHERE sha = ?"
+    );
+    $sth->execute($sha);
+    my $retval = int $sth->rows;
+    $sth->finish;
+    return $retval;
+}
+
+=head2 run($dbh)
+
+Runs against the current dbh without tracking.
+
+=cut
+
+sub run {
+    my ($self, $dbh) = @_;
+    $dbh->do($self->content); # not raw
+}
+
+=head2 apply($dbh)
+
+Applies the current file to the db in the current dbh.
+
+=cut
+
+sub apply {
+    my ($self, $dbh) = @_;
+    my $need_commit = _need_commit($dbh);
+    my $before = "";
+    my $after;
+    my $sha = $dbh->quote($self->sha);
+    my $path = $dbh->quote($self->path);
+    if ($self->is_applied($dbh)){
+        $after = "
+              UPDATE db_patches
+                     SET last_updated = now()
+               WHERE sha = $sha;
+        ";
+    } else {
+        $after = "
+           INSERT INTO db_patches (sha, path, last_updated)
+           VALUES ($sha, $path, now());
+        ";
+    }
+    if ($self->{no_transactions}){
+        $dbh->do($after);
+        $after = "";
+        $dbh->commit if $need_commit;
+    }
+    $dbh->do($self->content_wrapped($before, $after));
+    $dbh->commit if $need_commit;
+    warn "$dbh->state: $dbh->errstr";
+    $dbh->prepare("
+            INSERT INTO db_patch_log(when_applied, path, sha, sqlstate, error)
+            VALUES(now(), $path, $sha, ?, ?)
+    ")->execute($dbh->state, $dbh->errstr);
+    $dbh->commit if $need_commit;
+}
+
+sub _need_commit{
+    my ($dbh) = @_;
+    1; # todo, detect existing transactions and autocommit status
+}
+=head1 Package Functions
+
+=head2 init($dbh)
+
+Initializes the tracking system
+
+=cut
+
+sub init {
+    my ($dbh) = @_;
+    return 0 unless needs_init($dbh);
+    $dbh->do("
+    CREATE TABLE db_patch_log (
+       when_applied timestamp primary key,
+       path text NOT NULL,
+       sha text NOT NULL,
+       sqlstate text not null,
+       error text
+    );
+    CREATE TABLE db_patches (
+       sha text primary key,
+       path text not null,
+       last_updated timestamp not null
+    );
+    ");
+    return 1;
+}
+
+=head2 needs_init($dbh)
+
+Returns true if the tracking system needs to be initialized
+
+=cut
+
+sub needs_init {
+    my ($dbh) = @_;
+    my $rows = $dbh->prepare(
+       "select 1 from db_patches"
+    )->execute();
```

Comment:
What if RaiseError is true? Then when there is no patches table, this will die, I think?